crates/viewer/re_ui/src/list_item/custom_content.rs

```diff
@@ -0,0 +1,179 @@
+    ///
+    /// The closure will be called from within a [`egui::Ui`] with its maximum width set as per the
+    /// list item geometry. Note that this may differ from [`ContentContext::rect`] if a button is
+    /// set.
+    pub fn new(ui: impl FnOnce(&mut egui::Ui, &ContentContext<'_>) + 'a) -> Self {
+        Self {
+            ui: Box::new(ui),
+            desired_width: Default::default(),
+            button: None,
+        }
+    }
+
+    /// Set the desired width for the entire content.
+    #[inline]
+    pub fn with_desired_width(mut self, desired_width: DesiredWidth) -> Self {
+        self.desired_width = CustomContentDesiredWidth::DesiredWidth(desired_width);
+        self
+    }
+
+    /// Set the desired width based on the provided content width. If a button is set, its width
+    /// will be taken into account and added to the content width.
+    #[inline]
+    pub fn with_content_width(mut self, desired_content_width: f32) -> Self {
+        self.desired_width = CustomContentDesiredWidth::ContentWidth(desired_content_width);
+        self
+    }
+
+    /// Add a right-aligned [`super::ItemButton`].
+    ///
+    /// Note: for aesthetics, space is always reserved for the action button.
+    // TODO(#6191): accept multiple calls for this function for multiple actions.
+    #[inline]
+    pub fn button(mut self, button: impl super::ItemButton + 'a) -> Self {
+        // TODO(#6191): support multiple action buttons
+        assert!(
+            self.button.is_none(),
+            "Only one action button is supported right now"
+        );
+
+        self.button = Some(Box::new(button));
+        self
+    }
+
+    /// Helper to add an [`super::ItemActionButton`] to the right of the item.
+    ///
+    /// See [`Self::button`] for more information.
+    #[inline]
+    pub fn action_button(
+        self,
+        icon: &'static crate::icons::Icon,
+        on_click: impl FnOnce() + 'a,
+    ) -> Self {
+        self.action_button_with_enabled(icon, true, on_click)
+    }
+
+    /// Helper to add an enabled/disabled [`super::ItemActionButton`] to the right of the item.
+    ///
+    /// See [`Self::button`] for more information.
+    #[inline]
+    pub fn action_button_with_enabled(
+        self,
+        icon: &'static crate::icons::Icon,
+        enabled: bool,
+        on_click: impl FnOnce() + 'a,
+    ) -> Self {
+        self.button(super::ItemActionButton::new(icon, on_click).enabled(enabled))
+    }
+
+    /// Helper to add a [`super::ItemMenuButton`] to the right of the item.
+    ///
+    /// See [`Self::button`] for more information.
+    #[inline]
+    pub fn menu_button(
+        self,
+        icon: &'static crate::icons::Icon,
+        add_contents: impl FnOnce(&mut egui::Ui) + 'a,
+    ) -> Self {
+        self.button(super::ItemMenuButton::new(icon, add_contents))
+    }
+}
+
+impl ListItemContent for CustomContent<'_> {
+    fn ui(self: Box<Self>, ui: &mut egui::Ui, context: &ContentContext<'_>) {
```

Comment:
having a bit of a deja vu - isn't this button handling code already somewhere else? (can't tell without digging if that's actually the case and whether that's even a concern)

Reply:
Yes, I already have some kind of abstraction for button thingies (could be menu, could be action). Look for the `ItemButton` trait. I'm using that in `PropertyContent`, so I'm piggy backing on it.